### PR TITLE
feat(platform-browser): allow lazy-loading HammerJS

### PR DESCRIPTION
**Pull request by jelbourn**
Open for discussion. I made the loader Promise-based instead of Observable for simplicity here, but can change it if folks feel strongly.

### CODE REVIEW

**Comment by Toxicable** [comment]
This might be a reasonable case for `import()`?

**Comment by jelbourn** [comment]
@Toxicable the goal is to be agnostic to any particular loading implementation

**Comment by alfaproject** [comment]
I’ve been telling people to stop using Promises because Observables are more powerful. I think it would be good to be consistent in the framework side too

**Review by mhevery — APPROVED** [review]

File: packages/platform-browser/src/dom/events/hammer_gestures.ts [hunk 1]
```diff
@@ -58,6 +58,13 @@ const EVENT_NAMES = {
  */
 export const HAMMER_GESTURE_CONFIG = new InjectionToken<HammerGestureConfig>('HammerGestureConfig');
 
+
+/** Function that loads HammerJS, returning a promise that is resolved once HammerJs is loaded. */
+export type HammerLoader = (() => Promise<void>) | null;
```
Comment (vicb) [inline packages/platform-browser/src/dom/events/hammer_gestures.ts]: Is `| null` correct here ?

File: packages/platform-browser/src/dom/events/hammer_gestures.ts [hunk 2]
```diff
@@ -99,8 +106,8 @@ export class HammerGestureConfig {
 export class HammerGesturesPlugin extends EventManagerPlugin {
   constructor(
       @Inject(DOCUMENT) doc: any,
-      @Inject(HAMMER_GESTURE_CONFIG) private _config: HammerGestureConfig,
-      private console: Console) {
+      @Inject(HAMMER_GESTURE_CONFIG) private _config: HammerGestureConfig, private console: Console,
+      @Optional() @Inject(HAMMER_LOADER) private loader?: HammerLoader) {
```
Comment (vicb) [inline packages/platform-browser/src/dom/events/hammer_gestures.ts]: I think `| null` should be here instead ?

File: packages/platform-browser/src/dom/events/hammer_gestures.ts (hunk 1 above)
Comment (jelbourn) [reply packages/platform-browser/src/dom/events/hammer_gestures.ts]: I believe so; `null` is an acceptable value to `provide` for the token

File: packages/platform-browser/src/dom/events/hammer_gestures.ts (hunk 2 above)
Comment (jelbourn) [reply packages/platform-browser/src/dom/events/hammer_gestures.ts]: Per other comment: `null` is an acceptable value to provide for the token

File: packages/platform-browser/src/dom/events/hammer_gestures.ts [hunk 3]
```diff
@@ -58,6 +58,13 @@ const EVENT_NAMES = {
  */
 export const HAMMER_GESTURE_CONFIG = new InjectionToken<HammerGestureConfig>('HammerGestureConfig');
 
+
+/** Function that loads HammerJS, returning a promise that is resolved once HammerJs is loaded. */
+export type HammerLoader = (() => Promise<void>) | null;
+
+/** Injection token used to provide a {@link HammerLoader} to Angular. */
+export const HAMMER_LOADER = new InjectionToken<HammerLoader>('HammerLoader');
```
Comment (mhevery) [inline packages/platform-browser/src/dom/events/hammer_gestures.ts]: You need to add public docs and `@stable`here

Comment (petebacondarwin) [reply packages/platform-browser/src/dom/events/hammer_gestures.ts]: `@stable` is no longer needed nor permitted. It is implied from a lack of `@deprecated` and `@experimental` tags.

**Comment by cyrilletuzi** [comment]
@jelbourn Is there documentation somewhere to know how to use this feature? Thanks.

**Comment by jelbourn** [comment]
No documentation yet, but it boils down to providing a function as `HAMMER_LOADER`, where the function returns a `Promise` that is resolved once Hammer is available on the page. The function can load Hammer however you want.

**Pull request closed without merging**